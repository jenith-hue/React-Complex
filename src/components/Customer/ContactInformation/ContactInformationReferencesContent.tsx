import {
  RACButton,
  RACTable,
  RACTableCell,
  RACTableRow,
  RACTooltip,
  Typography,
} from '@rentacenter/racstrap';
import React from 'react';
import { useLocation } from 'react-router-dom';

import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { abbrev, formatPhoneNumber } from '../../../utils/utils';
import { CustomerLocationState } from '../../../types/types';
import { withContactStyles, WrappedComponentProps } from './withContactStyles';

export const ContactInformationReferencesContent_ = ({
  classes,
  onLogClicked,
}: WrappedComponentProps) => {
  const {
    customerDetails: { personalReferences: data },
    isLogActivityAllowed,
  } = useCustomerDetails();
  const location = useLocation<CustomerLocationState>();

  const renderTableHead = () => (
    <>
      <RACTableCell>Name</RACTableCell>
      <RACTableCell>Phone Number</RACTableCell>
      <RACTableCell>Relation</RACTableCell>
      <RACTableCell>Best Time to Call</RACTableCell>
      <RACTableCell># Contacts Today</RACTableCell>
      <RACTableCell>Last Call Result</RACTableCell>
      <RACTableCell align="center">Action</RACTableCell>
    </>
  );

  const customerId =
    location?.state?.customer?.customerId || location?.pathname?.split('/')[3];

  const renderTableContent = () => (
    <React.Fragment>
      {data?.map((dataObj) => (
        <RACTableRow
          backgroundColor="white"
          key={
            dataObj.personalReferenceId
              ? `PR${dataObj.personalReferenceId}`
              : `LR${dataObj.landlordReferenceId}`
          }
        >
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {dataObj.firstName
                ? `${dataObj.firstName} ${dataObj.lastName || ''}`
                : `${dataObj.landlordFirstName || ''} ${
                    dataObj.landlordLastName || ''
                  }`}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {formatPhoneNumber(
                dataObj.phoneNumber,
                dataObj.doNotCall === 'Y' || !isLogActivityAllowed
              )}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {dataObj.relationshipTypeDesc}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {dataObj.bestTimeToCallDesc}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography
              align="center"
              variant="body2"
              className={classes.tableContentColor}
            >
              {dataObj.communicationsToday}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography noWrap variant="body2" color="primary">
              <RACTooltip title={dataObj.lastCallResultDescription || ''}>
                <span>{abbrev(dataObj.lastCallResultDescription || '')}</span>
              </RACTooltip>
            </Typography>
          </RACTableCell>
          <RACTableCell align="center">
            <RACButton
              variant="outlined"
              color="primary"
              className={classes.logButton}
              onClick={() =>
                onLogClicked(customerId, dataObj.phoneNumber?.split('.')[0])
              }
              disabled={
                !dataObj?.phoneNumber ||
                !isLogActivityAllowed ||
                dataObj.doNotCall === 'Y'
              }
            >
              Log
            </RACButton>
          </RACTableCell>
        </RACTableRow>
      ))}
    </React.Fragment>
  );
  return (
    <RACTable
      renderTableHead={renderTableHead}
      renderTableContent={renderTableContent}
    />
  );
};

export const ContactInformationReferencesContent = (props: any) =>
  withContactStyles(props)(ContactInformationReferencesContent_);
