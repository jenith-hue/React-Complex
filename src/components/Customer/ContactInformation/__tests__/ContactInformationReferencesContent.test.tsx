/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { ContactInformationReferencesContent } from '../ContactInformationReferencesContent';
import RouteData from 'react-router-dom';

interface PropWithChildren {
  children: JSX.Element;
}

interface MockRACTableProps {
  renderTableHead: () => JSX.Element;
  renderTableContent: () => JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  ...jest.requireActual('@rentacenter/racstrap'),
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACButton: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACButton" {...rest}>
      {children}
    </div>
  ),
  RACCard: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACCard" {...rest}>
      {children}
    </div>
  ),
  RACTableCell: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableCell" {...rest}>
      {children}
    </div>
  ),
  RACTableRow: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableRow" {...rest}>
      {children}
    </div>
  ),
  RACTable: ({ renderTableHead, renderTableContent }: MockRACTableProps) => (
    <div data-testid="RACTable">
      <div id="table-head">{renderTableHead()}</div>
      <div id="table-content">{renderTableContent()}</div>
    </div>
  ),
  RACTooltip: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTooltip" {...rest}>
      {children}
    </div>
  ),
  makeStyles: () => () => ({}),
}));

jest.mock(
  '../../../../context/CustomerDetails/CustomerDetailsProvider',
  () => ({
    useCustomerDetails: () => ({
      customerDetails: {
        personalReferences: [
          {
            personalReferenceId: '24849',
            firstName: 'Alan',
            lastName: 'Hamel',
            relationshipTypeDesc: 'Friend',
            relationLengthYears: null,
            phoneNumber: '5431783478',
            bestTimeToCallDesc: null,
            verifiedDate: null,
            active: 'Y',
            doNotCall: 'N',
            note: null,
          },
          {
            personalReferenceId: '248491',
            firstName: 'Jean',
            lastName: 'Valjean',
            relationshipTypeDesc: 'Friend',
            relationLengthYears: null,
            phoneNumber: '5431783478',
            bestTimeToCallDesc: 'Afternoon',
            verifiedDate: null,
            active: 'Y',
            doNotCall: 'N',
            note: 'Customer not available',
          },
        ],
      },
    }),
  })
);

// Notice that the below is required, even if there is an another
// jest.spyOn in the test.
// Otherwise, we'll get the following: TypeError: Cannot redefine property: useLocation
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '',
  }),
}));
describe('ContanctInformationReferencesContent', () => {
  it('renders correctly', () => {
    jest.spyOn(RouteData, 'useLocation').mockImplementationOnce(() => ({
      pathname: '/testroute',
      search: '',
      hash: '',
      state: {
        customer: {
          customerFirstName: 'First',
          customerLastName: 'Last',
        },
      },
    }));
    const tree = renderer.create(<ContactInformationReferencesContent />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
