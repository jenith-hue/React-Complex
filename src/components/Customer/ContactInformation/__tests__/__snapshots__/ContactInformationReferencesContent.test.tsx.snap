// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContanctInformationReferencesContent renders correctly 1`] = `
<div
  data-testid="RACTable"
>
  <div
    id="table-head"
  >
    <div
      data-testid="RACTableCell"
    >
      Name
    </div>
    <div
      data-testid="RACTableCell"
    >
      Phone Number
    </div>
    <div
      data-testid="RACTableCell"
    >
      Relation
    </div>
    <div
      data-testid="RACTableCell"
    >
      Best Time to Call
    </div>
    <div
      data-testid="RACTableCell"
    >
      # Contacts Today
    </div>
    <div
      data-testid="RACTableCell"
    >
      Last Call Result
    </div>
    <div
      align="center"
      data-testid="RACTableCell"
    >
      Action
    </div>
  </div>
  <div
    id="table-content"
  >
    <div
      backgroundColor="white"
      data-testid="RACTableRow"
    >
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Alan Hamel
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          (***) ***-****
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Friend
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          align="center"
          data-testid="Typography"
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="Typography"
          noWrap={true}
          variant="body2"
        >
          <div
            data-testid="RACTooltip"
            title=""
          >
            <span>
              
            </span>
          </div>
        </div>
      </div>
      <div
        align="center"
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="RACButton"
          disabled={true}
          onClick={[Function]}
          variant="outlined"
        >
          Log
        </div>
      </div>
    </div>
    <div
      backgroundColor="white"
      data-testid="RACTableRow"
    >
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Jean Valjean
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          (***) ***-****
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Friend
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Afternoon
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          align="center"
          data-testid="Typography"
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="Typography"
          noWrap={true}
          variant="body2"
        >
          <div
            data-testid="RACTooltip"
            title=""
          >
            <span>
              
            </span>
          </div>
        </div>
      </div>
      <div
        align="center"
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="RACButton"
          disabled={true}
          onClick={[Function]}
          variant="outlined"
        >
          Log
        </div>
      </div>
    </div>
  </div>
</div>
`;
